File: creational/dart/builder_structured/builders.dart
import 'interfaces.dart';
import 'products.dart';

class SimpleHouseBuilder extends HouseBuilder {
  final House _house = House();
  
  @override
  void buildFoundation() => _house.setFoundation('Concrete foundation');
  
  @override
  void buildStructure() => _house.setStructure('Wood frame structure');
  
  @override
  void buildRoof() => _house.setRoof('Asphalt shingle roof');
  
  @override
  void buildInterior() => _house.setInterior('Basic interior with drywall');
  
  @override
  void buildGarage() => _house.setGarage(false);
  
  @override
  void buildGarden() => _house.setGarden(false);
  
  @override
  void buildSwimmingPool() => _house.setSwimmingPool(false);
  
  @override
  House getResult() => _house;
}

class LuxuryHouseBuilder extends HouseBuilder {
  final House _house = House();
  
  @override
  void buildFoundation() => _house.setFoundation('Reinforced concrete foundation with basement');
  
  @override
  void buildStructure() => _house.setStructure('Steel frame structure with brick exterior');
  
  @override
  void buildRoof() => _house.setRoof('Spanish tile roof with solar panels');
  
  @override
  void buildInterior() => _house.setInterior('Luxury interior with marble floors and custom cabinetry');
  
  @override
  void buildGarage() => _house.setGarage(true);
  
  @override
  void buildGarden() => _house.setGarden(true);
  
  @override
  void buildSwimmingPool() => _house.setSwimmingPool(true);
  
  @override
  House getResult() => _house;
}

class ComputerBuilder {
  final Computer _computer = Computer();
  
  ComputerBuilder setCpu(String cpu) {
    _computer.setCpu(cpu);
    return this;
  }
  
  ComputerBuilder setGpu(String gpu) {
    _computer.setGpu(gpu);
    return this;
  }
  
  ComputerBuilder setRam(int ram) {
    _computer.setRam(ram);
    return this;
  }
  
  ComputerBuilder setStorage(int storage) {
    _computer.setStorage(storage);
    return this;
  }
  
  ComputerBuilder addWifi() {
    _computer.setWifi(true);
    return this;
  }
  
  ComputerBuilder addBluetooth() {
    _computer.setBluetooth(true);
    return this;
  }
  
  Computer build() {
    return _computer;
  }
}

class SQLQueryBuilder {
  String _query = '';
  
  SQLQueryBuilder select(String columns) {
    _query = 'SELECT $columns';
    return this;
  }
  
  SQLQueryBuilder from(String table) {
    _query += ' FROM $table';
    return this;
  }
  
  SQLQueryBuilder where(String condition) {
    _query += ' WHERE $condition';
    return this;
  }
  
  SQLQueryBuilder orderBy(String column, {String direction = 'ASC'}) {
    _query += ' ORDER BY $column $direction';
    return this;
  }
  
  SQLQueryBuilder limit(int count) {
    _query += ' LIMIT $count';
    return this;
  }
  
  String build() {
    return '$_query;';
  }
}

File: creational/typescript/builder_structured/builders.ts
import { HouseBuilder } from './interfaces';
import { House, Computer } from './products';

export class SimpleHouseBuilder implements HouseBuilder {
  private house: House = new House();
  
  buildFoundation(): void { this.house.setFoundation('Concrete foundation'); }
  buildStructure(): void { this.house.setStructure('Wood frame structure'); }
  buildRoof(): void { this.house.setRoof('Asphalt shingle roof'); }
  buildInterior(): void { this.house.setInterior('Basic interior with drywall'); }
  buildGarage(): void { this.house.setGarage(false); }
  buildGarden(): void { this.house.setGarden(false); }
  buildSwimmingPool(): void { this.house.setSwimmingPool(false); }
  getResult(): House { return this.house; }
}

export class LuxuryHouseBuilder implements HouseBuilder {
  private house: House = new House();
  
  buildFoundation(): void { this.house.setFoundation('Reinforced concrete foundation with basement'); }
  buildStructure(): void { this.house.setStructure('Steel frame structure with brick exterior'); }
  buildRoof(): void { this.house.setRoof('Spanish tile roof with solar panels'); }
  buildInterior(): void { this.house.setInterior('Luxury interior with marble floors and custom cabinetry'); }
  buildGarage(): void { this.house.setGarage(true); }
  buildGarden(): void { this.house.setGarden(true); }
  buildSwimmingPool(): void { this.house.setSwimmingPool(true); }
  getResult(): House { return this.house; }
}

export class ComputerBuilder {
  private computer: Computer = new Computer();
  
  setCpu(cpu: string): ComputerBuilder { this.computer.setCpu(cpu); return this; }
  setGpu(gpu: string): ComputerBuilder { this.computer.setGpu(gpu); return this; }
  setRam(ram: number): ComputerBuilder { this.computer.setRam(ram); return this; }
  setStorage(storage: number): ComputerBuilder { this.computer.setStorage(storage); return this; }
  addWifi(): ComputerBuilder { this.computer.setWifi(true); return this; }
  addBluetooth(): ComputerBuilder { this.computer.setBluetooth(true); return this; }
  build(): Computer { return this.computer; }
}

export class SQLQueryBuilder {
  private query: string = '';
  
  select(columns: string): SQLQueryBuilder { this.query = `SELECT ${columns}`; return this; }
  from(table: string): SQLQueryBuilder { this.query += ` FROM ${table}`; return this; }
  where(condition: string): SQLQueryBuilder { this.query += ` WHERE ${condition}`; return this; }
  orderBy(column: string, direction: string = 'ASC'): SQLQueryBuilder { this.query += ` ORDER BY ${column} ${direction}`; return this; }
  limit(count: number): SQLQueryBuilder { this.query += ` LIMIT ${count}`; return this; }
  build(): string { return `${this.query};`; }
}
